tools/nl/validator/vartest.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+import os
+from typing import Dict, List
+
+from absl import app
+from absl import flags
+import requests
+import utils
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('input_csv', 'data/bootstrap.csv',
+                    'Input bootstrap CSV file')
+flags.DEFINE_string('names_csv', 'data/llm_output.csv',
+                    'Input SV descriptions CSV file')
+flags.DEFINE_string('checkpoint_dir', 'checkpoint/', 'Generated files')
+flags.DEFINE_string('run_name', 'foo',
+                    'Unique name of the test, for continuation, etc.')
+
+CONFIG = 'sz=medium_ft'
+
+URL = 'http://localhost:6060/api/search_sv?' + CONFIG
+
+OUT_HEADER = [
+    'Query',
+    'SV',
+    'EmbeddingsSVRank',
+    'BelowThresholdSVRank',
+    'Exception',
+    'EmptyResult',
+    'DemoteSV',
+    'BelowThresholdDemoteSV',
+]
+
+CHECKPOINT_INTERVAL = 100
+
+
+@dataclass
+class Context:
+  rows: List[Dict]
+  counters: Dict
+  bootstrap: Dict
+  input_csv: str
+  names_csv: str
+  output_csv: str
+  counters_json: str
+  out_header: List[str]
+
+
+def init_result(q, sv):
+  return {
+      'Query': q,
+      'SV': sv,
+      'Exception': '',
+      'EmptyResult': '',
+      'EmbeddingsSVRank': '',
+      'BelowThresholdSVRank': '',
+      'DemoteSV': [],
+      'BelowThresholdDemoteSV': [],
+  }
+
+
+def init_counters():
+  counters = {}
+  for c in [
+      'Total', 'Exception', 'EmptyResult', 'WrongPlace', 'EmbeddingsSVRank_0',
+      'EmbeddingsSVRank_1', 'EmbeddingsSVRank_2', 'EmbeddingsSVRank_3',
+      'EmbeddingsSVRank_4', 'EmbeddingsSVRank_5', 'EmbeddingsSVRank_6-14',
+      'EmbeddingsSVRank_15+', 'EmbeddingsSVRank_INF', 'BelowThresholdSVRank_x',
+      'BelowThresholdSVRank_INF'
+  ]:
+    counters[c] = 0
+  return counters
+
+
+def compose_query(v):
+  # clean up some stuff
+  v = v.lower()
+  if v.startswith('population: '):
+    v = v.replace('population: ', 'number of people who are ')
+  return v
+
+
+def query(sv, sv_name, counters):
```

Comment:
Could you add a couple of comments about the expected outcome here. Seems like this returns a dict with lists of SVs to demote etc?

Reply:
Also, note: when we add to the finetuning, we have to create pairs of sentences. So in an ideal case, we'd need the sentences that were matched to (in addition to the SV dcids). For example, demoting SV_i will mean we somehow need to look which SV_i sentence was matched that needs to be demoted when pairing with the query sentence.

Reply:
Added comments and finetuning data like we discussed.